server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java

```diff
@@ -362,9 +374,9 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         if (pathUnmatch != null) {
             builder.field("path_unmatch", pathUnmatch);
         }
-        if (xcontentFieldType != null) {
-            builder.field("match_mapping_type", xcontentFieldType);
-        } else if (match == null && pathMatch == null) {
+        if (xContentFieldTypes.length == 1) {
+            builder.field("match_mapping_type", xContentFieldTypes[0]);
+        } else if (xContentFieldTypes.length > 1 && match == null && pathMatch == null) {
```

Comment:
maybe this deserves a comment: empty array means the user did not specify it, 1 means a specific type was specified, > 1 means we expanded it to the corresponding types (depending on whether mapping or runtime is specified) while the user provided `*`

Reply:
I added a comment